lib/SILOptimizer/Utils/LexicalDestroyFolding.cpp

```diff
@@ -0,0 +1,687 @@
+  ///
+  /// returns false if the move_value [lexical] instruction was added already
+  ///         true if the move_value [lexical] instruction was added just now
+  bool createMove();
+
+  /// Combine the matched instruction sequence
+  ///
+  ///   apply %fn(%copy)
+  ///   end_borrow %lifetime
+  ///   destroy_value %instance
+  ///
+  /// into
+  ///
+  ///   end_borrow %lifetime
+  ///   apply %fn(%move)
+  ///
+  /// This is done in three steps:
+  /// (1) rewrite the apply (and delete the copy_value that is fed to it if
+  ///     possible)
+  /// (2) hoist the end_borrow
+  /// (3) delete the destroy_value
+  void fold(Match, SmallVectorImpl<int> const &rewritableArgumentIndices);
+
+  // Update SSA to reflect that fact that %move and %original_borrowee are two
+  // defs for the "same value".
+  void updateSSA();
+
+  /// The number of total uses of the simple extended borrow scope defined by
+  /// %introducer used as arguments in the specified apply instruction.
+  ///
+  /// Just does a lookup into extendedSimpleUseCountPerApplySite which is
+  /// calculated during findUsers.
+  size_t extendedSimpleUseCount(ApplyInst *ai) const;
+
+  /// Whether the specified value is could be replaced with %original_borrowee.
+  ///
+  /// In order to be, it has to be in the simple extended def of
+  /// %original_borrowee.
+  bool canReplaceValueWithBorrowee(SILValue value) const;
+
+  /// Find the arguments in the specified apply that could be rewritten.
+  void rewritableArgumentIndicesForApply(ApplySite,
+                                         SmallVectorImpl<int> &indices) const;
+
+private:
+  // State calculated over the course of the run.
+
+  // The instructions that are users of the simple extended borrow scope.
+  SmallPtrSet<SILInstruction *, 16> users;
+  // The operands that are uses of the original borrowee.
+  //
+  // Populated in findDominatedTransitiveUsesOfOriginalBorrowee.
+  SmallVector<Operand *, 16> originalBorroweeUses;
+  // Per apply site, the number of owned, simple extended uses of the borrow.
+  llvm::SmallDenseMap<ApplySite, size_t> extendedSimpleUseCountPerApplySite;
+  // The move_value instruction that was added during the run.
+  MoveValueInst *mvi = nullptr;
+  /// The sequences of instructions which are candidates to be folded.
+  ///
+  /// The list consists of triples:
+  /// - the match, the instruction sequences themselves
+  /// - whether the sequence is viable, as determined by
+  /// isViableInstructionSequence
+  /// - the argument indices which can and must be rewritten
+  llvm::SmallVector<std::tuple<Match, bool, SmallVector<int, 2>>, 4> candidates;
+};
+
+//===----------------------------------------------------------------------===//
+//                             MARK: Driver
+//===----------------------------------------------------------------------===//
+
+/// Perform the folding.
+///
+/// Returns whether any change was made.
+bool LexicalDestroyFolding::run() {
+  if (!findCandidates())
```

Comment:
> Ideally, the `LexicalDestroyFolding` class is not needed at all.

I think the specific suggestion was reasonable. When you can break apart an algorithm functionally into stages with distinct input and output, that's fantastic. But let's not get carried away. 

Passing around shared mutable state by reference makes the code less functional. To make the code more functional, you need to encapsulate all of the shared mutable state with the same lifetime into a single class. That allows a single stage of the algorithm to be decomposed into methods that each do one functional thing.

When an algorithm passes around shared mutable state in multiple data structures, it does several bad non-functional things:

- The functions no longer have a functional API. It's impossible to distinguish the function input/output from the shared mutable state.

- The shared mutable state is now accessed via many different local variables making it much harder to understand where it is accessed!

- It strongly discourages breaking an algorithm up into multiple functions. The code becomes very difficult to refactor.
  
- To understand an algorithm, the first thing you need to see are central data structures that are *not* localized to a single function. If those aren't encapsulated at the right level in a central class, then you need to read the entire implementation first.

Reply:
Because `candidates` gets written by two functions, it's now local to `LexicalDestroyFolding::run`.  The other fields are remaining as members variables of `LexicalDestroyFolding` because they're written once and then immutable.  It's documented when those member variables get defined.

Reply:
ok, so that's the opposite of what I was trying to say. I think Erik's specific suggestion to make 'users' and 'candidates' local was fine.

The general advice of "don't use class members to define state because that isn't functional" was wrong. Defining "local" variables and passing them by non-const reference to multiple functions is not functional either. It actually hides the mutable shared state which is even worse, and it doesn't establish any connection between the code that shares mutable state. Those variables are not really local at all. They are hidden globals that assume different name in each function.

```
void implementation() {
  ...
  DataStructure data1
  DataStructure data2
  helper1(data1, data2) // pass by ref
 .... way later...
  helper2(data1, data2) // pass by ref, I also modify shared state
 ... 
  helper3(data1, data2)
}
```

It's not a problem in your small run() function, but it becomes indecipherable very quickly. You can find lots of examples in our code base where several data structures are defined locally, then all passed individually into multiple functions. Outright terrible design.

Reply:
> "don't use class members to define state because that isn't functional"

That's not what I wanted to say. It makes sense to encapsulate multiple variables into structs/classes, because passing dozens of small variables to functions is not good for readability either.
Also, often it makes sense to have things like caches, etc. as global state. In this example I would keep `dominanceTree` and `deleter` as class members. Often it's a tradeoff and subject to personal taste.

But I would say that the algorithm in the `run` function, where results are _produced_ by some called functions and _used_ by other called functions, is clearly a case where those values should be local variables.
Documenting those "side-effects" is only the second best choice. The best comments are the comments which are not needed, because the behavior is obvious from the source code.

Reply:
For the record, I think we're all in pretty strong agreement.